fedimint-api/src/module/mod.rs

```diff
@@ -248,6 +250,132 @@ pub trait ModuleGen {
     fn validate_config(&self, identity: &PeerId, config: ServerModuleConfig) -> anyhow::Result<()>;
 }
 
+dyn_newtype_define!(
+    #[derive(Clone)]
+    pub DynModuleGen(Arc<IModuleGen>)
+);
+
+/// Module Generation trait with associated types
+///
+/// Needs to be implemented by module generation type
+///
+/// For examples, take a look at one of the `MintConfigGenerator`, `WalletConfigGenerator`, or
+/// `LightningConfigGenerator` structs.
+#[async_trait]
+pub trait ModuleGen: Debug + Sized {
+    const KIND: ModuleKind;
+
+    type Decoder: Decoder;
+
+    fn module_kind(&self) -> ModuleKind {
+        Self::KIND
+    }
+
+    fn decoder(&self) -> Self::Decoder;
+
+    /// Initialize the [`DynServerModule`] instance from its config
+    async fn init(
+        &self,
+        cfg: ServerModuleConfig,
+        db: Database,
+        env: &BTreeMap<OsString, OsString>,
+        task_group: &mut TaskGroup,
+    ) -> anyhow::Result<DynServerModule>;
+
+    fn trusted_dealer_gen(
+        &self,
+        peers: &[PeerId],
+        params: &ConfigGenParams,
+    ) -> BTreeMap<PeerId, ServerModuleConfig>;
+
+    async fn distributed_gen(
+        &self,
+        connections: &MuxPeerConnections<ModuleInstanceId, DkgPeerMsg>,
+        our_id: &PeerId,
+        module_id: ModuleInstanceId,
+        peers: &[PeerId],
+        params: &ConfigGenParams,
+        task_group: &mut TaskGroup,
+    ) -> anyhow::Result<Cancellable<ServerModuleConfig>>;
+
+    fn to_client_config(&self, config: ServerModuleConfig) -> anyhow::Result<ClientModuleConfig>;
+
+    fn to_client_config_from_consensus_value(
+        &self,
+        config: serde_json::Value,
+    ) -> anyhow::Result<ClientModuleConfig>;
+
+    fn validate_config(&self, identity: &PeerId, config: ServerModuleConfig) -> anyhow::Result<()>;
+}
+
+#[async_trait]
+impl<T> IModuleGen for T
+where
+    T: ModuleGen + 'static + Sync,
+{
+    fn decoder(&self) -> DynDecoder {
+        DynDecoder::from_typed(ModuleGen::decoder(self))
+    }
+
+    fn module_kind(&self) -> ModuleKind {
```

Comment:
same here

Reply:
I think this does though. otherwise `module_kind` wouldn't be object safe and I couldn't do `<Self as ModuleGen>`